clients/src/main/java/org/apache/kafka/clients/consumer/internals/MetadataSnapshot.java

```diff
@@ -0,0 +1,111 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE
+ * file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file
+ * to You under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ * <p/>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.apache.kafka.clients.consumer.internals;
+
+import org.apache.kafka.common.utils.Utils;
+
+import java.security.MessageDigest;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+
+/**
+ * Represents a point-in-time view of the topic metadata corresponding to a
+ * consumer's metadata.
+ *
+ * NOT thread-safe
+ */
+public class MetadataSnapshot {
+
+    private final MessageDigest digest;
+    private final Set<String> localSubscribedTopics;
+    private final Set<String> groupSubscribedTopics;
+    private final SortedMap<String, Integer> metadata; // topic -> # partitions
+
+    public MetadataSnapshot(
+            Set<String> localSubscribedTopics,
+            Set<String> groupSubscribedTopics,
+            SortedMap<String, Integer> metadata) {
+        this.localSubscribedTopics = localSubscribedTopics;
+        this.groupSubscribedTopics = groupSubscribedTopics;
+        this.metadata = metadata;
+
+        try {
+            digest = MessageDigest.getInstance("SHA-256");
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Get the number of partitions per topic
+     * @return Map from topic to number of partitions
+     */
+    public SortedMap<String, Integer> topicMetadata() {
+        return metadata;
+    }
+
+    /**
+     * Get the topics the local consumer has subscribed to
+     * @return A set of the local subscribed topics
+     */
+    public Set<String> localSubscibedTopics() {
+        return localSubscribedTopics;
+    }
+
+    /**
+     * Get the topics the group has collectively subscribed to
+     * @return A set of the group subscribed topics
+     */
+    public Set<String> groupSubscribedTopics() {
+        return groupSubscribedTopics;
+    }
+
+    public byte[] hash() {
```

Comment:
You might want to rename this for clarity. I saw `HashSets` being passed into the `MetadataSnapshot` constructor in `ConsumerGroupController` and since I knew `hash()` is called on that, I was worried the `HashSets` might have been a mistake. They're not, and that makes sense given the protocol we came up with, but the current method names aren't clear about what's being hashed. My first thought was `metadataHash()`, but that doesn't exactly help... maybe `topicMetadataHash()`?